File: yak-gl-api/src/main/java/net/yakclient/opengl/api/render/GLRenderingData.java
package net.yakclient.opengl.api.render;

import net.yakclient.opengl.util.*;

import java.nio.DoubleBuffer;
import java.nio.FloatBuffer;

/**
 * GLRendering data is considered actual data that needs to be
 * passed to opengl. It could for example be anything in a buffer
 * or a texture bound to opengl. It maintains data like vertice
 * count.
 * <p>
 * {@code GLRenderingContext} however doesnt contain what we consider
 * data but contains settings for opengl to process this data with.
 *
 * @author Durgan McBroom
 * @see VerticeRenderingContext
 * @see Aggregation
 * @see YakTexture
 */
public class GLRenderingData {
    private final VerticeAggregation vertices;
    private final ColorAggregation colors;
    private final NormalAggregation normals;
    private final TexAggregation texs;

    private final YakTexture texture;

    private final int verticeCount;

    public GLRenderingData(VerticeAggregation vertices, ColorAggregation colors, NormalAggregation normals, TexAggregation texs, YakTexture texture, int verticeCount) {
        this.vertices = vertices;
        this.colors = colors;
        this.normals = normals;
        this.texs = texs;
        this.texture = texture;
        this.verticeCount = verticeCount;
    }

    public GLRenderingData(RenderingDataBuilder builder) {
        this(builder.vertices, builder.colors, builder.normals, builder.texs, builder.texture, builder.verticeCount);
    }

    public DoubleBuffer getVertices() {
        return this.vertices.asBuf(this.vertices.createBuf());
    }

    public DoubleBuffer getNormals() {
        return this.normals.asBuf(this.normals.createBuf());
    }

    public FloatBuffer getColors() {
        return this.colors.asBuf(this.colors.createBuf());
    }

    public FloatBuffer getTexs() {
        return this.texs.asBuf(this.texs.createBuf());
    }

    public YakTexture getTexture() {
        return texture;
    }

    public boolean hasVertices() {
        return !this.vertices.isEmpty();
    }

    public boolean hasColors() {
        return !this.colors.isEmpty();
    }

    public boolean hasNormals() {
        return !this.normals.isEmpty();
    }

    public boolean hasTexs() {
        return !this.texs.isEmpty();
    }

    public int getVerticeCount() {
        return verticeCount;
    }

    public static RenderingDataBuilder create() {
        return new RenderingDataBuilder();
    }

    public static RenderingDataBuilder create(YakTexture texture) {
        return new RenderingDataBuilder(texture);
    }

    public static class RenderingDataBuilder {
        private final VerticeAggregation vertices;
        private final ColorAggregation colors;
        private final NormalAggregation normals;
        private final TexAggregation texs;

        private final YakTexture texture;

        private int verticeCount = 0;


        public RenderingDataBuilder() {
            this.vertices = new VerticeAggregation();
            this.colors = new ColorAggregation();
            this.normals = new NormalAggregation();
            this.texs = new TexAggregation();
            this.texture = new VacantTexture();
        }

        public RenderingDataBuilder(YakTexture texture) {
            this.vertices = new VerticeAggregation();
            this.colors = new ColorAggregation();
            this.normals = new NormalAggregation();
            this.texs = new TexAggregation();
            this.texture = texture;
        }

        public RenderingDataBuilder addColor3f(float r, float g, float b) {
            this.colors.add(r, g, b);
            return this;
        }

        public RenderingDataBuilder addColor4f(float r, float g, float b, float alpha) {
            this.colors.add(r, b, b, alpha);
            return this;
        }

        public RenderingDataBuilder addColors(ColorAggregation aggregation) {
            this.colors.combine(aggregation);
            return this;
        }

        public RenderingDataBuilder addVertice2d(double x, double y) {
            this.verticeCount++;
            this.vertices.add(x, y);
            return this;
        }

        public RenderingDataBuilder addVertice3d(double x, double y, double z) {
            this.verticeCount++;
            this.vertices.add(x, y, z);
            return this;
        }

        public RenderingDataBuilder addVertices(VerticeAggregation aggregation) {
            this.verticeCount += aggregation.size();
            this.vertices.combine(aggregation);
            return this;
        }

        public RenderingDataBuilder addNormal(double x, double y, double z) {
            this.normals.add(x, y, z);
            return this;
        }

        public RenderingDataBuilder addNormals(NormalAggregation aggregation) {
            this.normals.combine(aggregation);
            return this;
        }

        public RenderingDataBuilder addTexVertice(float t, float s) {
            this.texs.add(t, s);
            return this;
        }

        public RenderingDataBuilder addTexs(TexAggregation aggregation) {
            this.texs.combine(aggregation);
            return this;
        }

        public GLRenderingData build() {
            return new GLRenderingData(this);
        }
    }
}

File: api/src/main/kotlin/net/yakclient/graphics/api/render/GLRenderingData.kt
package net.yakclient.graphics.api.render

import net.yakclient.graphics.util.VerticeAggregation
import net.yakclient.graphics.util.ColorAggregation
import net.yakclient.graphics.util.NormalAggregation
import net.yakclient.graphics.util.TexAggregation
import net.yakclient.graphics.util.YakTexture
import java.nio.DoubleBuffer
import java.nio.FloatBuffer
import net.yakclient.graphics.util.VacantTexture

/**
 * GLRendering data is considered actual data that needs to be
 * passed to opengl. It could for example be anything in a buffer
 * or a texture bound to opengl. It maintains data like vertice
 * count.
 *
 *
 * `GLRenderingContext` however doesnt contain what we consider
 * data but contains settings for opengl to process this data with.
 *
 * @author Durgan McBroom
 * @see VerticeRenderingContext
 *
 * @see Aggregation
 *
 * @see YakTexture
 */
class GLRenderingData(
    public val vertices: VerticeAggregation,
    private val colors: ColorAggregation,
    private val normals: NormalAggregation,
    private val texs: TexAggregation,
    val texture: YakTexture,
    val verticeCount: Int
) {

    constructor(builder: RenderingDataBuilder) : this(builder.vertices,
        builder.colors,
        builder.normals,
        builder.texs,
        builder.texture,
        builder.verticeCount) {
    }



//    fun getVertices(): DoubleBuffer {
//        return vertices.asBuf(vertices.createBuf())
//    }
//
//    fun getNormals(): DoubleBuffer {
//        return normals.asBuf(normals.createBuf())
//    }
//
//    fun getColors(): FloatBuffer {
//        return colors.asBuf(colors.createBuf())
//    }
//
//    fun getTexs(): FloatBuffer {
//        return texs.asBuf(texs.createBuf())
//    }

    public fun hasVertices(): Boolean {
        return !vertices.isEmpty
    }

    public fun hasColors(): Boolean {
        return !colors.isEmpty
    }

    public fun hasNormals(): Boolean {
        return !normals.isEmpty
    }

    public fun hasTexs(): Boolean {
        return !texs.isEmpty
    }

    class RenderingDataBuilder {
        val vertices: VerticeAggregation
        val colors: ColorAggregation
        val normals: NormalAggregation
        val texs: TexAggregation
        val texture: YakTexture
        var verticeCount = 0

        constructor() {
            vertices = VerticeAggregation()
            colors = ColorAggregation()
            normals = NormalAggregation()
            texs = TexAggregation()
            texture = VacantTexture()
        }

        constructor(texture: YakTexture) {
            vertices = VerticeAggregation()
            colors = ColorAggregation()
            normals = NormalAggregation()
            texs = TexAggregation()
            this.texture = texture
        }

        fun addColor3f(r: Float, g: Float, b: Float): RenderingDataBuilder {
            colors.add(r, g, b)
            return this
        }

        fun addColor4f(r: Float, g: Float, b: Float, alpha: Float): RenderingDataBuilder {
            colors.add(r, b, b, alpha)
            return this
        }

        fun addColors(aggregation: ColorAggregation): RenderingDataBuilder {
            colors.combine(aggregation)
            return this
        }

        fun addVertice2d(x: Double, y: Double): RenderingDataBuilder {
            verticeCount++
            vertices.add(x, y)
            return this
        }

        fun addVertice3d(x: Double, y: Double, z: Double): RenderingDataBuilder {
            verticeCount++
            vertices.add(x, y, z)
            return this
        }

        fun addVertices(aggregation: VerticeAggregation): RenderingDataBuilder {
            verticeCount += aggregation.size()
            vertices.combine(aggregation)
            return this
        }

        fun addNormal(x: Double, y: Double, z: Double): RenderingDataBuilder {
            normals.add(x, y, z)
            return this
        }

        fun addNormals(aggregation: NormalAggregation): RenderingDataBuilder {
            normals.combine(aggregation)
            return this
        }

        fun addTexVertice(t: Float, s: Float): RenderingDataBuilder {
            texs.add(t, s)
            return this
        }

        fun addTexs(aggregation: TexAggregation): RenderingDataBuilder {
            texs.combine(aggregation)
            return this
        }

        fun build(): GLRenderingData {
            return GLRenderingData(this)
        }
    }

    companion object {
        fun create(): RenderingDataBuilder {
            return RenderingDataBuilder()
        }

        @JvmStatic
        fun create(texture: YakTexture): RenderingDataBuilder {
            return RenderingDataBuilder(texture)
        }
    }
}